doc/rtd/topics/code_review.rst

```diff
@@ -0,0 +1,170 @@
+*******************
+Code Review Process
+*******************
+
+In order to manage incoming pull requests effectively, and to ensure
+that community members (like you!) submitting them understand what to
+expect, cloud-init has a documented code review process.
+
+The language used is intentionally formal, in a bid to avoid ambiguity,
+but the *experience* of working through the code review process should
+not be, so please don't be put off!  If you have any questions about
+this process, please do ask them in ``#cloud-init`` on the Freenode IRC
+network.
+
+Goals
+=====
+
+This process has the following goals:
+
+* Ensure that cloud-init community members are receiving code reviews
+  in a timely fashion
+* Minimize the time the cloud-init core developers have to spend on the
+  parts of the code review process which aren't actual code review
+
+Roles
+=====
+
+For ease of understanding, we will refer to two different roles
+throughout the process:
+
+Proposer
+   The developer proposing a pull request (hereafter known as a PR).
+   This could be a core developer or community member.  (This is likely
+   the role you will be taking in this process.)
+
+the Committers
+   The group of cloud-init core developers.
+
+   In this process doc, cloud-init core developers are treated as a
+   bloc of interchangeable people.  This is a simplification, as in
+   reality we will need to manage transfer of reviews, dismissal of
+   stale reviews, parallel reviews from multiple committers, etc.  If
+   we find that handling these on an ad-hoc basis is causing problems,
+   then we should revisit and expand this process to include them.
+
+   (If the **Proposer** is a core developer, then they are not
+   considered part of this group for the purposes of this process.)
+
+Opening Phase
+=============
+
+In this phase, the **Proposer** is responsible for all actions. They
+should get a PR into good enough shape that it is worth **the
+Committers** spending time reviewing it.  Specifically, they are
```

Comment:
if it's not in good enough shape yet, the proposer could open the pr as WIP, or draft

Reply:
Hmm, this is an interesting point. We need to strike a balance here: we don't want to encourage people to submit half-baked PRs in the hopes that we will pick it up and do the work for them, but we also do want people (like you) who are working on chunkier changes to not have to wait until they have a full change to propose for us to get eyes on it.

My gut feeling is that the people in the latter camp will all be involved in development on a regular basis, so we can discuss exceptions to the process in IRC or bug comments. This perhaps means that we don't need to capture this into the process itself? What do you think?

(Also, as a side note: not everyone can apply/remove labels, so we can't use them as a self-serve process step. I had to revise an earlier draft of this process once I realised that.)

Reply:
"good enough" has to be defined. After all early on it was stated that we have this document to remove ambiguity. The definition of what "good enough" means should occur prior to using the term.

Further there is the case where one needs advice/feedback before proceeding. Such conversations are best captured in the conversation in a PR, thus "good enough" may actually not apply. ANd to folow the finger wagging tone of the document ;) it better be worth "the committers" time to provide feedback on something that is WIP to help unblock someone that is working on code and got stuck.

Lastly this can be written without wagging the finger, which is the impression the current chosen wording provides.

Core team members, as well as all other contributors should be here to help. If that help is to provide feedback to someone that might be new to the project and whose initial attempt may not be what one would ordinarily consider to be in "good shape" then IMHO it is still worth while to help that person along. The wording here doesn't really make me feel welcome and encouraged to contribute to the project.

I think it is worth to define various categories here for example:
- Needs feedback on approach
- Needs help, I am stuck
- Ready for merge review (and one criteria for this would be all tests pass)

Then we can define the process of adding this categorization into the comments since not everyone can set labels. Something like this also clarifies the state of the PR. For example "Needs help, I am stuck" should have a very specific question such as "this test is failing and I don't know why. I've looked here and there but cannot figure it out" or "I do not understand the structure of the test setup in XYZ and thus cannot add tests for my new code here, can you please explain"

I think this section needs more clarity and differentiation and yes the wording should aim to be inclusive and friendly.